sys/usb/usbus/hid/hid.c

```diff
@@ -0,0 +1,228 @@
+
+    DEBUG("hid_init: %d %d \n", report_desc_size, report_desc[0]);
+    usbus_register_event_handler(usbus, &hid->handler_ctrl);
+}
+
+static void _init(usbus_t *usbus, usbus_handler_t *handler)
+{
+    DEBUG("USB_HID: initialization\n");
+    usbus_hid_device_t *hid = (usbus_hid_device_t *)handler;
+
+    hid->tx_ready.handler = _handle_tx_ready;
+
+    hid->hid_descr.next = NULL;
+    hid->hid_descr.funcs = &_hid_descriptor;
+    hid->hid_descr.arg = hid;
+
+    /*
+    Configure Interface as USB_HID interface, choosing NONE for subclass and
+    protocol in order to represent a generic I/O device
+    */
+    hid->iface.class = USB_CLASS_HID;
+    hid->iface.subclass = USB_HID_SUBCLASS_NONE;
+    hid->iface.protocol = USB_HID_PROTOCOL_NONE;
+    hid->iface.descr_gen = &hid->hid_descr;
+    hid->iface.handler = handler;
+
+    usbus_endpoint_t *ep = usbus_add_endpoint(usbus, &hid->iface,
+                                              USB_EP_TYPE_INTERRUPT,
+                                              USB_EP_DIR_IN,
+                                              CONFIG_USBUS_HID_INTERRUPT_EP_SIZE);
+
+    /* interrupt endpoint polling rate */
+    ep->interval = 0x05;
+    usbus_enable_endpoint(ep);
+
+    ep = usbus_add_endpoint(usbus, &hid->iface,
+                            USB_EP_TYPE_INTERRUPT, USB_EP_DIR_OUT,
+                            CONFIG_USBUS_HID_INTERRUPT_EP_SIZE);
+
+    /* interrupt endpoint polling rate */
+    ep->interval = 0x05;
+    usbus_enable_endpoint(ep);
+
+    /* signal that INTERRUPT OUT ready to receive data */
+    usbdev_ep_ready(ep->ep, 0);
+
+    usbus_add_interface(usbus, &hid->iface);
+}
+
+static void _event_handler(usbus_t *usbus, usbus_handler_t *handler,
+                           usbus_event_usb_t event)
+{
+    (void)usbus;
+    (void)handler;
+
+    switch (event) {
+        default:
+            DEBUG("USB HID unhandled event: 0x%x\n", event);
+            break;
+    }
+}
+
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                            usbus_control_request_state_t state,
+                            usb_setup_t *setup)
+{
+    usbus_hid_device_t *hid = (usbus_hid_device_t *)handler;
+
+    DEBUG("USB_HID: request: %d type: %d value: %d length: %d state: %d \n",
+          setup->request, setup->type, setup->value >> 8, setup->length, state);
+
+    /* Requests defined in USB HID 1.11 spec section 7 */
+    switch (setup->request) {
+        case USB_SETUP_REQ_GET_DESCRIPTOR: {
+            uint8_t desc_type = setup->value >> 8;
+            if (desc_type == USB_HID_DESCR_REPORT) {
+                usbus_control_slicer_put_bytes(usbus, hid->report_desc,
+                                               hid->report_desc_size);
+            }
+            else if (setup->value >> 8 == USB_HID_DESCR_HID) {
+                _gen_hid_descriptor(usbus, NULL);
+            }
+            break;
+        }
+        case USB_HID_REQUEST_GET_REPORT:
+            break;
```

Comment:
Just for my understanding: if we were to implement a keyboard, this would be the part where the host would query the key events?

Reply:
It can be, yes. The HID spec states: "Get_Report request allows the host to receive a report via the Control pipe.".

Receiving a report via the Control pipe is not the normal case though. Normally, the key events would be polled by the host
via the Interrupt pipe.